src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/TrainVariantAnnotationModel.java

```diff
@@ -0,0 +1,260 @@
+import org.broadinstitute.hellbender.utils.clustering.BayesianGaussianMixtureModeller;
+import org.broadinstitute.hellbender.utils.io.IOUtils;
+import org.broadinstitute.hellbender.utils.python.PythonScriptExecutor;
+import org.broadinstitute.hellbender.utils.runtime.ProcessOutput;
+import picard.cmdline.programgroups.VariantFilteringProgramGroup;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * TODO
+ */
+@CommandLineProgramProperties(
+        // TODO
+        summary = "",
+        oneLineSummary = "",
+        programGroup = VariantFilteringProgramGroup.class
+)
+@DocumentedFeature
+public final class TrainVariantAnnotationModel extends CommandLineProgram {
+
+    private static final String SCORER_SER_SUFFIX = ".scorer.ser";
+    private static final String BGMM_HDF5_SUFFIX = ".bgmm.hdf5";
+
+    @Argument(
+            fullName = "annotations-hdf5",
+            doc = "HDF5 file containing annotations extracted with ExtractAnnotations.")
+    private File inputAnnotationsHDF5File;
+
+    @Argument(
+            fullName = "python-script",
+            optional = true)
+    private File pythonScriptFile;
+
+    @Argument(
+            fullName = "hyperparameters-json",
+            doc = "JSON file containing hyperparameters.")
+    private File hyperparametersJSONFile;
+
+    @Argument(
+            fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME,
+            shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME,
+            doc = "Output prefix.")
+    private String outputPrefix;
+
+    // TODO get rid of this argument, it's only needed for tranches
+    @Argument(
+            fullName = "mode",
+            shortName = "mode",
+            doc = "Variant type to extract")
+    public VariantTypeMode mode = VariantTypeMode.SNP;
+
+    /**
+     * Add truth sensitivity slices through the call set at the given values. The default values are 100.0, 99.9, 99.0, and 90.0
+     * which will result in 4 estimated tranches in the final call set: the full set of calls (100% sensitivity at the accessible
+     * sites in the truth set), a 99.9% truth sensitivity tranche, along with progressively smaller tranches at 99% and 90%.
+     * Note: You must pass in each tranche as a separate value (e.g. -tranche 100.0 -tranche 99.9).
+     */
+    @Argument(
+            fullName = "truth-sensitivity-tranche",
+            shortName = "tranche",
+            doc = "The levels of truth sensitivity at which to slice the data. (in percent, that is 1.0 for 1 percent)",
+            optional = true)
+    private List<Double> truthSensitivityTranches = new ArrayList<>(Arrays.asList(100.0, 99.9, 99.0, 90.0));
+
+    @Override
+    protected Object doWork() {
```

Comment:
Break this up into smaller pieces.

Reply:
Extracted a few methods, but the result is still a bit procedural and lengthy (100 lines) now that it includes positive-negative as well. However, a decent fraction of those lines is from self-documenting logging. It's also a bit hard to break up more because many intermediate outputs are reused---extracted methods would have lengthy and confusing signatures.

That said, I think it's worth leaving further refactoring as a possible exercise for a future developer and have added documentation to this effect. Expanding this into a full, iterative positive-unlabeled approach (we might consider the VQSR-style positive-negative method as the first iteration of such a full approach) might also be a meatier but worthwhile endeavor.